File: 19. Remove Nth Node From End of List.py
# https://leetcode.com/problems/remove-nth-node-from-end-of-list/

# Given the head of a linked list, remove the nth node from the end of the list
# and return its head.

# Follow up: Could you do this in one pass?

###############################################################################

# linked list -> nth node from end -> slow and fast ptrs

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def removeNthFromEnd(self, head: ListNode, n: int) -> ListNode:
        dummy = ListNode()
        dummy.next = head
        slow, fast = dummy, dummy
        
        # move fast n time
        for _ in range(n):
            fast = fast.next
        
        # move fast and slow together
        while fast.next: # stop at fast = tail, slow.next = desired node
            slow = slow.next
            fast = fast.next
        
        slow.next = slow.next.next # skip the desired node
        return dummy.next

File: 19. Remove Nth Node From End of List.cpp
// https://leetcode.com/problems/remove-nth-node-from-end-of-list/

// Given the head of a linked list, remove the nth node from the end of the list
// and return its head.

////////////////////////////////////////////////////////////////////////////////

// use slow and fast ptrs
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        ListNode *dummy = new ListNode();
        dummy->next = head;
        ListNode *slow = dummy, *fast = dummy;
        
        while (n-- > 0) fast = fast->next; // move fast n times;
        while (fast->next) { // move slow and fast until fast is at end
            slow = slow->next;
            fast = fast->next;
        }
        slow->next = slow->next->next; // skip the desire node
        return dummy->next;
    }
};
